### PR TITLE
Implement sealed type parameters

### PR DESCRIPTION
**Pull request by odersky**
Abolish restriction that box/unbox cannot be done with `cap`.

Instead, do a deep check for `cap`

 - in type arguments for `sealed` type variables
 - in the types of mutable vars
 - in the type of `try` under the `saferExceptions` language import

The `caps.unsafe.{unsafeBox, unsafeUnbox, unsafeBoxFunArg}` methods are not longer
needed and are deprecated. Instead there is an annotation `annotation.unchecked.uncheckedCaptures` that
can be added to a mutable variable to turn off checking its type.

These changes in behavior are introduced in 3.3. Older source versions
still support the old behavior (which corresponds to the paper).

So to run examples in the paper as described there, they need a `-source 3.2` or a
language import.

Based on #17377

### CODE REVIEW

**Comment by odersky** [comment]
Hi Yichen,

Can you review, and in particular double check that the changes to tests in [ce3b6d2](https://github.com/lampepfl/dotty/pull/17422/commits/ce3b6d24768c402b25d875597c3241d3ff48bfc9) all make sense? I.e. that there's nothing slipping through that should be caught.

File: tests/pending/neg/cc-depfun.scala
Comment (Linyxus) [inline tests/pending/neg/cc-depfun.scala:1]: This issue can be reproduced before this PR. It seems that the inconsistency is a result of `expandAbbreviations`:
https://github.com/lampepfl/dotty/blob/9ed9839e533db22c69bdb1e552ef6eeaacecc333/compiler/src/dotty/tools/dotc/cc/Setup.scala#L286-L296

File: tests/pending/neg/cc-depfun.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import language.experimental.captureChecking
```
Comment (Linyxus) [inline tests/pending/neg/cc-depfun.scala:1]: This inconsistency can be reproduced before this PR. It seems that this is a result of `expandAbbreviations`:
https://github.com/lampepfl/dotty/blob/9ed9839e533db22c69bdb1e552ef6eeaacecc333/compiler/src/dotty/tools/dotc/cc/Setup.scala#L286-L296

File: tests/pending/neg/cc-depfun.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+trait Cap { def use(): Unit }
+
+def main() = {
+  val f: (io: Cap^) -> () -> Unit =
+    io => () => io.use()  // error
+
+  val g: (Cap^) -> () -> Unit =
+    io => () => io.use()  // error
+}
```
Comment (Linyxus) [inline tests/pending/neg/cc-depfun.scala:14]: This inconsistency can be reproduced before this PR. It seems that this is a result of `expandAbbreviations`:
https://github.com/lampepfl/dotty/blob/9ed9839e533db22c69bdb1e552ef6eeaacecc333/compiler/src/dotty/tools/dotc/cc/Setup.scala#L286-L296

**Review by Linyxus — APPROVED** [review]
LGTM!

**Pull request merged**